server/server/src/billing/billing_service.rs

```diff
@@ -1,194 +1,361 @@
+            )
+                .await
+                .map_err(|err| internal!("{:?}", err))?;
+
+            sub_profile.data_cap = FREE_TIER_USAGE_SIZE;
+            info!("Successfully canceled stripe subscription. public_key: {:?}", context.public_key);
 
-        if current_time - user_info.last_in_payment_flow < millis_between_payment_flows {
-            info!(
-                "User is already in payment flow, or this request is too soon after a failed one. public_key: {}",
-                keys::stringify_public_key(public_key)
-            );
-            return Err(Abort(ClientError(SwitchAccountTierError::ConcurrentRequestsAreTooSoon)));
         }
+    }
+
+    reset_billing_lock(&context.public_key, &mut con, &mut sub_profile).await?;
 
-        user_info.last_in_payment_flow = current_time;
+    Ok(CancelSubscriptionResponse {})
+}
+
+pub async fn upgrade_account_stripe(
+    context: RequestContext<'_, UpgradeAccountStripeRequest>,
+) -> Result<UpgradeAccountStripeResponse, ServerError<UpgradeAccountStripeError>> {
+    let (request, server_state) = (&context.request, context.server_state);
 
-        pipe.json_set(stripe_user_info(public_key), &user_info)?;
+    info!("Attempting to switch account tier of {:?} to premium", context.public_key);
 
-        Ok(&mut pipe)
+    let mut con = server_state.index_db_pool.get().await?;
+
+    let mut sub_profile = lock_billing_workflow(
+        &context.public_key,
+        &mut con,
+        server_state.config.stripe.millis_between_user_payment_flows,
+    )
+    .await
+    .map_err(|err| match err {
+        ClientError(LockBillingWorkflowError::ExistingRequestPending) => {
+            ClientError(UpgradeAccountStripeError::ExistingRequestPending)
+        }
+        InternalError(msg) => InternalError(msg),
+    })?;
+
+    if sub_profile.data_cap == PREMIUM_TIER_USAGE_SIZE {
+        return Err(ClientError(UpgradeAccountStripeError::AlreadyPremium));
+    }
+
+    let maybe_user_info = sub_profile.billing_platform.and_then(|info| {
+        if let BillingPlatform::Stripe(stripe_info) = info {
+            Some(stripe_info)
+        } else {
+            None
+        }
     });
-    return_if_error!(tx_result);
+
+    let (user_info, new_data_cap) = create_subscription(
+        server_state,
+        &mut con,
+        &context.public_key,
+        &request.account_tier,
+        maybe_user_info,
+    )
+    .await?;
+
+    sub_profile.billing_platform = Some(BillingPlatform::Stripe(user_info));
+    sub_profile.data_cap = new_data_cap;
+
+    reset_billing_lock(&context.public_key, &mut con, &mut sub_profile).await?;
 
     info!(
-        "User successfully entered payment flow. public_key: {}",
-        keys::stringify_public_key(public_key)
+        "Successfully switched the account tier of {:?} from free to premium.",
+        context.public_key
     );
 
-    Ok(user_info)
+    Ok(UpgradeAccountStripeResponse {})
 }
 
 async fn create_subscription(
```

Comment:
should probably not have this general a name anymore. maybe should be located in a stripe specific file?

Reply:
Not too sure what you mean, can you explain?